shipyard-controller/handler/shipyardcontroller.go

```diff
@@ -241,122 +277,92 @@ func (sc *shipyardController) appendLatestCommitIDToEvent(eventScope models.Even
+		return err
 	}
 
-	if !wasTriggered {
-		log.Infof("The received %s event with keptn context %s is not accociated with a task that was previously triggered",
-			eventScope.EventType, eventScope.KeptnContext)
+	// now check if the number of .started events matches the number of finished events - if yes, that means were done
+	// note: this should also work with multiple replicas because the `AppendTaskEvent` updates the list of events and returns the resulting state
+	// atomically, so ONLY the thread that appended the last event to reach the completion state of the task will get the state required for further proceeding with the task sequence
+	if !updatedSequenceExecution.Status.CurrentTask.IsFinished() {
 		return nil
 	}
 
-	common.LockServiceInStageOfProject(eventScope.Project, eventScope.Stage, eventScope.Service+":taskFinisher")
-	defer common.UnlockServiceInStageOfProject(eventScope.Project, eventScope.Stage, eventScope.Service+":taskFinisher")
+	result, status := updatedSequenceExecution.CompleteCurrentTask()
+
+	eventScope.Result = result
+	eventScope.Status = status
 
-	startedEvents, err := sc.eventRepo.GetStartedEventsForTriggeredID(*eventScope)
+	triggeredEventType, err := keptnv2.ReplaceEventTypeKind(eventScope.EventType, string(common.TriggeredEvent))
 	if err != nil {
-		return fmt.Errorf("error while retrieving matching '.started' event for event %s with triggered ID %s: %w", eventScope.WrappedEvent.ID, eventScope.TriggeredID, err)
-	}
-	if len(startedEvents) == 0 {
-		return fmt.Errorf("no matching '.started' event found for event %s with triggered ID %s", eventScope.WrappedEvent.ID, eventScope.TriggeredID)
+		return err
 	}
 
-	err = sc.eventRepo.InsertEvent(eventScope.Project, eventScope.WrappedEvent, common.FinishedEvent)
+	triggeredEvents, err := sc.eventRepo.GetEventsWithRetry(eventScope.Project, common.EventFilter{Type: triggeredEventType, ID: &eventScope.TriggeredID}, common.TriggeredEvent, maxRepoReadRetries)
 	if err != nil {
-		log.Errorf("unable to store %s event: %v ", eventScope.EventType, err.Error())
+		return fmt.Errorf("unable to retrieve associated task '.triggered' event with ID %s: %w", eventScope.TriggeredID, err)
+	}
+	if len(triggeredEvents) == 0 {
+		return fmt.Errorf("no matching task '.triggered' event found for event %s with triggered ID %s", eventScope.WrappedEvent.ID, eventScope.TriggeredID)
 	}
 
-	err = sc.eventRepo.DeleteEvent(eventScope.Project, startedEvents[0].ID, common.StartedEvent)
+	// the '.triggered' event can now be removed
+	err = sc.eventRepo.DeleteEvent(eventScope.Project, triggeredEvents[0].ID, common.TriggeredEvent)
 	if err != nil {
-		return fmt.Errorf("unable to delete associated task '.started' event with ID %s: %w", startedEvents[0].ID, err)
+		return fmt.Errorf("unable to delete associated task '.triggered' event with ID %s: %w", eventScope.TriggeredID, err)
 	}
 
-	// if this was the last '.started' event
```

Comment:
now we just use `sequenceExecution.Status.CurrentTask.IsFinished()` instead of retrieving and comparing events from various collections :-)